gu-client/src/async.rs

```diff
@@ -0,0 +1,297 @@
+#![allow(dead_code)]
+
+use actix_web::{client, HttpMessage};
+use bytes::Bytes;
+use error::Error;
+use futures::stream::Stream;
+use futures::{future, Future};
+use gu_net::rpc::peer::PeerInfo;
+use std::str;
+use std::sync::Arc;
+
+/// Hub session information.
+#[derive(Clone, Serialize, Deserialize, Debug, Default, Builder)]
+#[builder(pattern = "owned", setter(into))]
+pub struct SessionInfo {
+    name: String,
+    environment: String,
+}
+
+/// Peer session information.
+#[derive(Clone, Serialize, Deserialize, Debug, Default, Builder)]
+#[builder(pattern = "owned", setter(into))]
+pub struct PeerSessionInfo {
+    image: PeerSessionImage,
+    environment: String,
+}
+
+/// Peer session image.
+#[derive(Clone, Serialize, Deserialize, Debug, Default)]
+pub struct PeerSessionImage {
+    url: String,
+    hash: String,
+}
+
+/// Connection to a single hub.
+#[derive(Clone, Debug)]
+pub struct Driver {
+    driver_inner: Arc<DriverInner>,
+}
+
+#[derive(Debug)]
+struct DriverInner {
+    url: String,
+}
+
+impl Driver {
+    /// creates a driver from a given address:port, e.g. 127.0.0.1:61621
+    pub fn from_addr<T: Into<String>>(addr: T) -> Driver {
+        Driver {
+            driver_inner: Arc::new(DriverInner {
+                url: format!("http://{}/", addr.into()),
+            }),
+        }
+    }
+    /// creates a new hub session
+    /// assumption: url is /sessions, hub session is in the response body
+    pub fn new_session(
+        &self,
+        session_info_builder: SessionInfoBuilder,
+    ) -> impl Future<Item = HubSession, Error = Error> {
+        let sessions_url = format!("{}sessions", self.driver_inner.url);
+        let session_info = match session_info_builder.build() {
+            Ok(r) => r,
+            _ => return future::Either::A(future::err(Error::InvalidHubSessionParameters)),
```

Comment:
I may be missing something, but why don't you just use `map_err` and `and_then`?

Reply:
I just checked again and you are not missing anything: it can be replaced with `map_err` and `and_then`, but it is less readable, as there is a second `match(...)` later and the code is even more indented.